node/src/test/scala/com/wavesplatform/mining/LightNodeBlockFieldsTest.scala

```diff
@@ -0,0 +1,113 @@
+      AddrWithBalance.enoughBalances(defaultSigner, secondSigner)
+    ) { d =>
+      withMiner(
+        d.blockchain,
+        d.testTime,
+        d.settings.copy(minerSettings = d.settings.minerSettings.copy(quorum = 0, minMicroBlockAge = 0.seconds)),
+        verify = false,
+        timeDrift = Int.MaxValue
+      ) { case (miner, append) =>
+        val microBlockMiner = new MicroBlockMinerImpl(
+          _ => (),
+          null,
+          d.blockchainUpdater,
+          d.utxPool,
+          d.settings.minerSettings,
+          miner.minerScheduler,
+          miner.appenderScheduler,
+          Observable.empty,
+          identity
+        )
+        val challenger = new BlockChallengerImpl(
+          d.blockchain,
+          new DefaultChannelGroup(GlobalEventExecutor.INSTANCE),
+          d.wallet,
+          d.settings,
+          d.testTime,
+          d.posSelector,
+          b => Task.now(append(b)),
+          timeDrift = Int.MaxValue
+        ) {
+          override def pickBestAccount(accounts: Seq[(SeedKeyPair, Long)]): Either[GenericError, (SeedKeyPair, Long)] = Right((defaultSigner, 0))
+        }
+        def block(height: Int) = d.blocksApi.blockAtHeight(height).get._1.header
+        def appendBlock()      = append(miner.forgeBlock(defaultSigner).explicitGet()._1).explicitGet()
+        def appendMicro() = {
+          d.utxPool.putIfNew(transfer()).resultE.explicitGet()
+          microBlockMiner.generateOneMicroBlockTask(defaultSigner, d.lastBlock, Unlimited, 0).runSyncUnsafe()
+        }
+        def challengeBlock() = {
+          val invalidBlock = d.createBlock(ProtoBlockVersion, Seq(), strictTime = true, stateHash = Some(Some(ByteStr.fill(DigestLength)(1))))
+          challenger.challengeBlock(invalidBlock, null).runSyncUnsafe()
+        }
+
+        appendBlock()
+        d.blockchain.height shouldBe 2
+        d.blockchain.isFeatureActivated(LightNode) shouldBe true
+        block(2).stateHash shouldBe None
+
+        appendMicro()
+        block(2).stateHash shouldBe None
+
+        challengeBlock()
+        d.blockchain.height shouldBe 3
+        block(3).stateHash shouldBe None
+        block(3).challengedHeader shouldBe None
+
+        (1 to 998).foreach(_ => appendBlock())
+        d.blockchain.height shouldBe 1001
+        block(1001).stateHash shouldBe None
+
+        appendMicro()
+        block(1001).stateHash shouldBe None
+
+        appendBlock()
+        d.blockchain.height shouldBe 1002
+        val hash1 = block(1002).stateHash
+        hash1 shouldBe defined
+
+        appendMicro()
+        val hash2 = block(1002).stateHash
+        hash2 shouldBe defined
+        hash2 should not be hash1
+
+        d.rollbackTo(1000)
+        challengeBlock()
+        block(1001).stateHash shouldBe None
+        block(1001).challengedHeader shouldBe None
+
+        challengeBlock()
+        block(1002).stateHash shouldBe defined
+        block(1002).challengedHeader shouldBe defined
```

Comment:
```suggestion
        (1 to 8).foreach(_ => appendBlock())
        d.blockchain.height shouldBe 11
        block(11).stateHash shouldBe None

        appendMicro()
        block(11).stateHash shouldBe None

        appendBlock()
        d.blockchain.height shouldBe 12
        val hash1 = block(12).stateHash
        hash1 shouldBe defined

        appendMicro()
        val hash2 = block(12).stateHash
        hash2 shouldBe defined
        hash2 should not be hash1

        d.rollbackTo(10)
        challengeBlock()
        block(11).stateHash shouldBe None
        block(11).challengedHeader shouldBe None

        challengeBlock()
        block(12).stateHash shouldBe defined
        block(12).challengedHeader shouldBe defined
```
Less blocks - faster tests :)

Reply:
Done